native/src/main/native/org/mozilla/jss/pkcs11/PK11KeyGenerator.c

```diff
@@ -282,53 +349,64 @@ Java_org_mozilla_jss_pkcs11_PK11KeyGenerator_generatePBE(
     }
     /* print_secitem(pwitem); */
 
-    mech = JSS_getPK11MechFromAlg(env, alg);
+    if(hashAlg==NULL) {
+    	mech = JSS_getPK11MechFromAlg(env, alg);
+    }
 
     if( mech == CKM_PBA_SHA1_WITH_SHA1_HMAC ) {
-
         /* special case, construct key by hand. Bug #336587 */
-
         skey = constructSHA1PBAKey(env, slot, pwitem, salt, iterationCount);
         if( skey==NULL ) {
             /* exception was thrown */
             goto finish;
         }
+    }
 
-    } else {
-
-        /* get the algorithm info */
-        oidTag = JSS_getOidTagFromAlg(env, alg);
-        PR_ASSERT(oidTag != SEC_OID_UNKNOWN);
-
-        SECOidTag encAlgOidTag = JSS_getOidTagFromAlg(env, encAlg);
-        PR_ASSERT(encAlgOidTag != SEC_OID_UNKNOWN);
-
-        /* create algid */
-        algid = PK11_CreatePBEV2AlgorithmID(
-            oidTag,
-            encAlgOidTag,
-            SEC_OID_HMAC_SHA1,
-            0,
-            iterationCount,
-            salt);
-
-        if( algid == NULL ) {
-            JSS_throwMsg(env, TOKEN_EXCEPTION,
-                    "Unable to process PBE parameters");
+    if( mech == CKM_SHA256 || mech == CKM_SHA384 || mech == CKM_SHA512  ) {
+    	skey = constructSHAPBEKey(env, mech, slot, pwitem, salt, iterationCount);
+        if( skey==NULL ) {
             goto finish;
         }
+    }
 
-        /* generate the key */
-        skey = PK11_PBEKeyGen(slot, algid, pwitem, PR_FALSE /*faulty3DES*/,
-                        NULL /*wincx*/);
-        if( skey == NULL ) {
-            JSS_throwMsg(env, TOKEN_EXCEPTION, "Failed to generate PBE key");
-            goto finish;
-        }
+    if( skey == NULL ) {
+    	/* get the algorithm info */
+		oidTag = JSS_getOidTagFromAlg(env, alg);
+		PR_ASSERT(oidTag != SEC_OID_UNKNOWN);
+
+		SECOidTag encAlgOidTag = JSS_getOidTagFromAlg(env, encAlg);
+		SECOidTag hashAlgOidTag = JSS_getOidTagFromAlg(env, hashAlg);
+		PR_ASSERT(encAlgOidTag != SEC_OID_UNKNOWN);
+		PR_ASSERT(hashAlgOidTag != SEC_OID_UNKNOWN);
+
+		/* create algid */
+		algid = PK11_CreatePBEV2AlgorithmID(
+			oidTag,
+			encAlgOidTag,
+			hashAlgOidTag,
+			0,
+			iterationCount,
+			salt);
+
+		if( algid == NULL ) {
+			JSS_throwMsg(env, TOKEN_EXCEPTION,
+					"Unable to process PBE parameters");
+			goto finish;
+		}
+
+		/* generate the key */
+		skey = PK11_PBEKeyGen(slot, algid, pwitem, PR_FALSE /*faulty3DES*/,
+						NULL /*wincx*/);
+		if( skey == NULL ) {
+			JSS_throwMsg(env, TOKEN_EXCEPTION, "Failed to generate PBE key");
+			goto finish;
+		}
     }
 
     /* wrap the key. This sets skey to NULL. */
-    keyObj = JSS_PK11_wrapSymKey(env, &skey);
+    if( skey!=NULL ) {
+        keyObj = JSS_PK11_wrapSymKey(env, &skey);
+    }
```

Comment:
Also one more quick thing I'm just putting at the end for convenience. Since we've done some significant work in the native code, I usually take it upon myself to run some memory analyzer such as Valgrind on a few of the major manual test cases to see if any memory leaks have escaped the review. Often it's easy to miss simple instances of failing to free memory or resources. If you want to use valgrind, just google valgrind quick start and just run your major pk12 test files through and see if it reports anything inside any of the code that was added. This is of course if you haven't done this sort of thing already.

Reply:
This is my first valgrind test. I have build the code with debug opetion and run a test with valgrind.
I am attaching the output for the memory check. It seems there are no memory leak with the c code modified here but let me know if I run it in a wrong way.

[valgrind.txt](https://github.com/dogtagpki/jss/files/10086815/valgrind.txt)

Reply:
Looks like you did it correctly. For convenience, instead of using > you can specify everything to go to a log file with this invocation:  valgrind --log-file="filename"

I didn't find any reports in the native files you modified. Anyway maybe try one more run with the --log-file switch to make sure we are seeing everything possible. If you see pretty much the same output, then we can call it good.

Reply:
This is definitely better because I can see the output of the program much easier. No problem with the modified file. This is the log file:
[valgrind.log](https://github.com/dogtagpki/jss/files/10122641/valgrind.log)